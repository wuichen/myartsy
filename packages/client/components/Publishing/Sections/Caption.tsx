import { garamond, unica } from "Assets/Fonts"
import { pMedia } from "Components/Helpers"
import { ArticleLayout, SectionLayout } from "Components/Publishing/Typings"
import React from "react"
import styled from "styled-components"

interface CaptionProps {
  caption: string
  color?: string
  index?: any
  layout?: ArticleLayout
  sectionLayout?: SectionLayout
}

interface FigcaptionProps {
  color?: string
  layout: ArticleLayout
  sectionLayout?: SectionLayout
}

export const Caption: React.SFC<CaptionProps> = props => {
  const { caption, children, color, layout, sectionLayout } = props

  const child = children || (
    <div dangerouslySetInnerHTML={{ __html: caption }} />
  )

  return (
    <CaptionContainer>
      <Figcaption layout={layout} sectionLayout={sectionLayout} color={color}>
        {child}
      </Figcaption>
    </CaptionContainer>
  )
}

export const CaptionContainer = styled.div`
  display: flex;
  justify-content: space-between;
  margin: 10px 0 10px 0;

  ${pMedia.xs`
    padding: 0px 10px;
  `};
`

// includes draft placeholder class for editable text in Writer
const Figcaption = styled.div<FigcaptionProps>`
  padding: ${props => (props.sectionLayout === "fillwidth" ? "0 10px;" : "0;")}
  width: 100%;
  word-break: break-word;

  & > p, p,
  .public-DraftEditorPlaceholder-root,
  .public-DraftStyleDefault-block {
    ${props => (props.layout === "classic" ? garamond("s15") : unica("s14"))}
    color: ${props =>
      props.color ? props.color : props.layout === "classic" ? "#666" : "#999"};
    margin: 0;
    display: inline;    
  }

  & > a, a {
    color: #999;
    &:hover {
      color: black;
    }
  }

  ${pMedia.xs`
    padding: 0px;
  `}
`
