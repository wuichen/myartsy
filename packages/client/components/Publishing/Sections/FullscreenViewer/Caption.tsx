import Colors from "Assets/Colors"
import { unica } from "Assets/Fonts"
import { pMedia } from "Components/Helpers"
import PropTypes from "prop-types"
import React from "react"
import styled, { StyledFunction } from "styled-components"
import { ArtworkCaption } from "../ArtworkCaption"

interface CaptionProps extends React.HTMLProps<HTMLDivElement> {
  section?: any
  total: number
  index: number
  open: boolean
}

interface CaptionOpenProps extends React.HTMLProps<HTMLDivElement> {
  open: boolean
}

export const Caption: React.SFC<CaptionProps> = props => {
  const { index, open, section, total } = props

  const isArtwork = section.type === "artwork"
  const indexText = `${index} of ${total}`

  return (
    <CaptionContainer>
      <CaptionTextContainer>
        <CaptionToggle open={open} />

        <CaptionText open={open}>
          {isArtwork ? (
            <ArtworkCaption artwork={section} isFullscreenCaption linked />
          ) : (
            <div
              dangerouslySetInnerHTML={{
                __html: section.caption,
              }}
            />
          )}
        </CaptionText>
      </CaptionTextContainer>

      <Index>{indexText}</Index>
    </CaptionContainer>
  )
}

const CaptionToggle: React.SFC<CaptionOpenProps> = (props, context) => {
  const toggleMessage = props.open ? "Hide" : "View Caption"

  return (
    <StyledCaptionToggle
      onClick={context.onToggleCaption}
      className="fullscreen-viewer__caption-toggle"
    >
      <span>{toggleMessage}</span>
    </StyledCaptionToggle>
  )
}

CaptionToggle.contextTypes = {
  onToggleCaption: PropTypes.func,
}

const StyledCaptionToggle = styled.div`
  display: none;
  ${pMedia.sm`
    ${unica("s14", "medium")}
    cursor: pointer;
    display: inline-block;
    span {
      border-bottom: 1px solid black;
    }
  `};
`

const CaptionDiv: StyledFunction<CaptionOpenProps> = styled.div

const CaptionText = CaptionDiv`
  ${unica("s16", "medium")}
  a {
    color: black;
  }
  ${props => pMedia.sm`
    ${unica("s14", "medium")}
    display: ${props.open ? "block" : "none"};
    margin-top: ${props.open ? "20px" : "0px"};
  `}
`

const Index = styled.div`
  margin-left: 20px;
  white-space: nowrap;
  ${unica("s16")};
  ${pMedia.sm`
    ${unica("s14")}
  `};
`

const CaptionContainer = styled.div`
  display: flex;
  justify-content: space-between;
  background-color: ${Colors.gray};
  padding: 30px 60px;

  p {
    margin: 0;
  }

  ${pMedia.sm`
    position: absolute;
    bottom: 0;
    width: 100%;
    max-width: 100vh;
    padding: 20px;
  `};
`

const CaptionTextContainer = styled.div`
  display: flex;
  flex-direction: column;
`
