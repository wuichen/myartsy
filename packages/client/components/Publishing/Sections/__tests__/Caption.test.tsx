import "jest-styled-components"
import React from "react"
import renderer from "react-test-renderer"
import { Caption } from "../Caption"

it("renders a saved caption properly", () => {
  const embed = renderer
    .create(<Caption caption="<p>This is a saved caption</p>" />)
    .toJSON()
  expect(embed).toMatchSnapshot()
})

it("renders a child even if a caption is provided", () => {
  const embed = renderer
    .create(
      <Caption caption="<p>This is a saved caption</p>">
        <div>Here is a passed child</div>
      </Caption>
    )
    .toJSON()
  expect(embed).toMatchSnapshot()
})
