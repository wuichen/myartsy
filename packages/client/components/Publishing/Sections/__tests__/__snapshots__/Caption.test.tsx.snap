// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a child even if a caption is provided 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 10px 0 10px 0;
}

.c1 {
  padding: 0;
  width: 100%;
  word-break: break-word;
}

.c1 > p,
.c1 p,
.c1 .public-DraftEditorPlaceholder-root,
.c1 .public-DraftStyleDefault-block {
  font-family: Unica77LLWebRegular,Arial,serif;
  -webkit-font-smoothing: antialiased;
  font-size: 14px;
  line-height: 1.4em;
  color: #999;
  margin: 0;
  display: inline;
}

.c1 > a,
.c1 a {
  color: #999;
}

.c1 > a:hover,
.c1 a:hover {
  color: black;
}

@media (max-width:600px) {
  .c0 {
    padding: 0px 10px;
  }
}

@media (max-width:600px) {
  .c1 {
    padding: 0px;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div>
      Here is a passed child
    </div>
  </div>
</div>
`;

exports[`renders a saved caption properly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 10px 0 10px 0;
}

.c1 {
  padding: 0;
  width: 100%;
  word-break: break-word;
}

.c1 > p,
.c1 p,
.c1 .public-DraftEditorPlaceholder-root,
.c1 .public-DraftStyleDefault-block {
  font-family: Unica77LLWebRegular,Arial,serif;
  -webkit-font-smoothing: antialiased;
  font-size: 14px;
  line-height: 1.4em;
  color: #999;
  margin: 0;
  display: inline;
}

.c1 > a,
.c1 a {
  color: #999;
}

.c1 > a:hover,
.c1 a:hover {
  color: black;
}

@media (max-width:600px) {
  .c0 {
    padding: 0px 10px;
  }
}

@media (max-width:600px) {
  .c1 {
    padding: 0px;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>This is a saved caption</p>",
        }
      }
    />
  </div>
</div>
`;
